### PR TITLE
ROC-6116 Add callback for initiate citizen payment

### PR DESCRIPTION
**Pull request by simad**
**Before creating a pull request make sure that:**

- [x] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/ROC-6117

### Change description ###

Adding an endpoint to initiate a citizen payment. This endpoint triggers an event and a ccd callback, in which we calculate fees and create a payment for a citizen.
This endpoint returns either the next url of the payment to frontend, or an error.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java [hunk 1]
```diff
@@ -184,6 +186,11 @@ public Claim getClaimByExternalId(String externalId, User user) {
         return caseRepository.getClaimsByState(claimState, user);
     }
 
+    public InitiatePaymentResponse initiatePayment(String authorisation, String submitterId, InitiatePaymentRequest request) {
+        User user = userService.getUser(authorisation);
+        return caseRepository.initiatePayment(user, submitterId, request);
+    }
```
Comment (andrewwa-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java]: FMI: Will the payhub client be invoked before or after this?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/CoreCaseDataService.java [hunk 2]
```diff
@@ -768,6 +773,57 @@ public void saveDirectionsQuestionnaireDeadline(Long caseId, LocalDate dqDeadlin
...
+                caseDataContent,
+                user.isRepresented()
+            );
+
+            return InitiatePaymentResponse.builder()
+                .nextUrl(caseDetails.getData().get(PAYMENT_NEXT_URL).toString())
+                .build();
+
+        } catch (Exception exception) {
+            System.out.println("ERROR" + exception.getMessage());
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/CoreCaseDataService.java]: haha that was when I was debugging stuff with @ckalista . Will remove :D

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/Payment.java [hunk 3]
```diff
@@ -27,39 +27,23 @@
     @JsonProperty("date_created")
     private final String dateCreated;
     private final String status;
+    @NotBlank
+    @JsonProperty("next_url")
```
[the remark this answers is not in the document]
Comment (simad) [reply domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/Payment.java]: 👍

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java [hunk 4]
```diff
@@ -184,6 +186,11 @@ public Claim getClaimByExternalId(String externalId, User user) {
         return caseRepository.getClaimsByState(claimState, user);
     }
 
+    public InitiatePaymentResponse initiatePayment(String authorisation, String submitterId, InitiatePaymentRequest request) {
```
[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java]: Yes.

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ClaimServiceTest.java [hunk 5]
```diff
@@ -490,6 +491,20 @@ public void updateStateShouldCallCaseRepository() {
...
+        when(userService.getUser(eq(AUTHORISATION))).thenReturn(USER);
+        when(userService.getUserDetails(AUTHORISATION)).thenReturn(VALID_CLAIMANT);
+        when(caseRepository.getClaimByExternalId(eq(EXTERNAL_ID), any()))
+            .thenReturn(Optional.of(claim));
+
+        InitiatePaymentRequest initiatePaymentRequest = InitiatePaymentRequest.builder().build();
+
+        claimService.initiatePayment(AUTHORISATION, "submitterId", initiatePaymentRequest);
+
+        verify(caseRepository, once()).initiatePayment(USER, "submitterId", initiatePaymentRequest);
```
[the remark this answers is not in the document]
Comment (simad) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/services/ClaimServiceTest.java]: yes, I know it's redundant. It's called copy paste 🙌

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/Payment.java [hunk 6]
```diff
@@ -27,39 +27,23 @@
...
 
     public Payment(
         String id,
         BigDecimal amount,
         String reference,
         String dateCreated,
-        String status
-    ) {
+        String status,
+        String nextUrl) {
```
[the remark this answers is not in the document]
Comment (simad) [reply domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/Payment.java]: I have been not really following the de facto standard 😬

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java (hunk 1 above)
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java]: after, the pay hub client will be invoked in the callback, ie. https://github.com/hmcts/cmc-claim-store/pull/1124/files#diff-ef12ef8695ae4d3fe705d8258e9b382bR39

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/ioc/InitiatePaymentRequest.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.UUID;
+import javax.validation.Valid;
+import javax.validation.constraints.NotNull;
+
+@Data
+@Builder
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class InitiatePaymentRequest {
+    @Valid
+    private final UUID externalId;
```
Comment (sabahirfan) [inline domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/ioc/InitiatePaymentRequest.java]: can this be null?

Comment (simad) [reply domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/ioc/InitiatePaymentRequest.java]: nope - nicely spotted

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java [hunk 8]
```diff
@@ -175,12 +178,15 @@ private void linkToCaseWithoutRevoking(
     }
 
     private List<Claim> getAllCasesBy(User user, ImmutableMap<String, String> searchString) {
-        return extractClaims(searchAll(user, searchString));
+        return extractClaims(asStream(searchAll(user, searchString))
+            .filter(isAwaitingCitizenState.negate())
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: I am thinking, wouldn't this be better if we pass this as filter as search criteria to CCD. Probably need to give it a try whether it works. 
Please refer here:
https://github.com/hmcts/cmc-claim-store/blob/2ab561410fa921b96376e01c05d85dca285f5823/src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java#L337

Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: 👍 I'll take a look and see if it works. Thanks!

**Pull request closed without merging**